File: WindowsRegular Expressions/4975sourcecode/chapter4/4-15/recipe4-15.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

public class Recipe
{
	private static string month29 = "0?2/(0?[1-9]|[12][0-9])";	
	private static string month30 = "(0?[469]|11)/(0?[1-9]|[12][0-9]|30)";
	private static string month31 = "(0?[13578]|1[02])/(0?[1-9]|[12][0-9]|3[01])";
	private static Regex _Regex = new Regex(String.Format("^({0}|{1}|{2})",
														  month29,
														  month30,
														  month31
														  ) + @"/\d{4}$");
		
	public void Run(string fileName)
	{
		String line;
		int lineNbr = 0;
		using (StreamReader sr = new StreamReader(fileName))
		{
			while(null != (line = sr.ReadLine()))
			{
				lineNbr++;
				if (_Regex.IsMatch(line))
				{
					Console.WriteLine("Found match '{0}' at line {1}", 
						line, 
						lineNbr);
				}
			}
		}
	}	

	public static void Main( string[] args )
	{
		Recipe r = new Recipe();
		r.Run(args[0]);
	}
}

File: WindowsRegular Expressions/4975sourcecode/chapter4/4-15/recipe4-15.vb
Imports System
Imports System.IO
Imports System.Text.RegularExpressions
Public Class Recipe

	Private Const month29 = "0?2/(0?[1-9]|[12][0-9])"
	Private Const month30 = "(0?[469]|11)/(0?[1-9]|[12][0-9]|30)"
	Private Const month31 = "(0?[13578]|1[02])/(0?[1-9]|[12][0-9]|3[01])"
	Private Shared _Regex As Regex = New Regex(String.Format("^({0}|{1}|{2})", _
														  month29, _
														  month30, _
														  month31 _
														  ) & "/\d{4}$")

	Public Sub Run(ByVal fileName As String)
		Dim line As String
		Dim lineNbr As Integer = 0
		Dim sr As StreamReader = File.OpenText(fileName)
		line = sr.ReadLine
		While Not line Is Nothing
			lineNbr = lineNbr + 1
			If _Regex.IsMatch(line) Then
				Console.WriteLine("Found match '{0}' at line {1}", _
					line, _
					lineNbr)
			End If
			line = sr.ReadLine
		End While
		sr.Close()
	End Sub

	Public Shared Sub Main(ByVal args As String())
		Dim r As Recipe = New Recipe
		r.Run(args(0))
	End Sub

End Class
